### PR TITLE
FEATURE: Enforce 2FA

### LINKED ISSUE #5: FEATURE: Warn of missing 2FA / Enforce setting up 2FA
My customer asked whether it can be made possible to warn of a missing 2FA token when a user logs in.
Or even force a redirect to setting it up as long as it hasn't been done yet.

The might be a customer budget for this, then I could provide a PR.
If not I just leave it here as an idea :)

### PR DESCRIPTION
**Pull request by JamesAlias**
Provide a settings option to enforce the use of 2FA for all neos users.

## TODO
- [x] Documentation!
> All issues below fixed
- [x] **When** on "enforced" 2FA setup screen and I can change the URL to `/neos` **then** I get to the content view and like this I am able to circumvent the enforced setup of a second factor
- [x] **When** second factor is set up & I correctly log in with username and password **then** I enter an infinite redirect loop `login->two-factor-login->two-factor-login->two-factor-login->....`
- [x] **When** I have no session cookie on `/neos/login` **then** no new session is created
- [x] **When** I have no session **then** I get redirected to `Sandstorm\NeosTwoFactorAuthentication\Controller\LoginController` and I get an exception because the the controller can not read `account` from the `AuthToken`

closes #5

### CODE REVIEW

File: Classes/Http/Middleware/SecondFactorMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+            );
+            return $next->handle($request);
+        }
+
+        if (
+            $this->secondFactorRepository->isEnabledForAccount($account)
+            && $authenticationStatus === AuthenticationStatus::AUTHENTICATION_NEEDED
+        ) {
+            // WHY: We use the request URI as part of state. This prevents the middleware to enter a redirect loop.
+            $isAskingForOTP = str_ends_with($request->getUri()->getPath(), 'neos/two-factor-login');
```
Comment (JamesAlias) [inline Classes/Http/Middleware/SecondFactorMiddleware.php]: make URI constant

File: Classes/Domain/AuthenticationStatus.php [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace Sandstorm\NeosTwoFactorAuthentication\Domain;
+
+enum AuthenticationStatus: string
```
Comment (JamesAlias) [inline Classes/Domain/AuthenticationStatus.php]: Do not use enum

File: Configuration/Settings.yaml [hunk 3]
```diff
@@ -22,7 +22,7 @@ Neos:
     http:
       middlewares:
         'secondFactorRedirectMiddleware':
-          middleware: 'Sandstorm\NeosTwoFactorAuthentication\Http\Middleware\SecondFactorRedirectMiddleware'
+          middleware: 'Sandstorm\NeosTwoFactorAuthentication\Http\Middleware\SecondFactorMiddleware'
```
Comment (JamesAlias) [inline Configuration/Settings.yaml:25]: position? `after x`
x = security/authentication/sessionMiddleware

File: Configuration/Settings.yaml [hunk 4]
```diff
@@ -31,34 +31,11 @@ Neos:
...
-#                patternOptions:
-#                  controllerObjectNamePattern: 'Neos\Neos\Service\.*'
-#            entryPoint: WebRedirect
-#            entryPointOptions:
-#              routeValues:
-#                '@package': Sandstorm.NeosTwoFactorAuthentication
-#                '@controller': Login
-#                '@action': askForSecondFactor
-#                '@format': html
+                  controllerObjectNamePattern: 'Sandstorm\NeosTwoFactorAuthentication\Controller\(AuthenticationController|LoginController|BackendController)'
```
Comment (JamesAlias) [inline Configuration/Settings.yaml]: remove `AuthenticationController`

**Comment by Benjamin-K** [comment]
This also closes #9.

File: Configuration/Settings.yaml (hunk 3 above)
Comment (JamesAlias) [reply Configuration/Settings.yaml:25]: `position: 'after securityEntryPoint'`

**Comment by JamesAlias** [comment]
This also resolves #6

**Pull request merged**